<a href="https://colab.research.google.com/github/araujoroberts/Proyectos/blob/main/Ecuaci%C3%B3n_de_%C3%93rbita_del_Meteorito.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Planteamiento del problema

Un meteorito se dirige a la casa blanca y nos interesa determinar su trayectoria, para esto necesitamos condiciones iniciales de posición y velocidad. Tomemos las siguientes:

$$
\vec{r}_0=(\ 0.50000000\ ,\ -6.5000000\ ,\ 4.5000000\ ) \ \ \ [\ m\ \times\ 10^6\ ]
$$

$$
\vec{r}_f=(\ 1.1123098\ ,\ -4.8319904\ ,\ 4.0005346\ ) \ \ \ [\ m\ \times\ 10^6\ ]
$$

_____________________________________________________________________

Inicializamos los vectores $\ \vec{r}_0\ $ y $\ \vec{r}_f\ $ 

In [1]:
import numpy as np
import math

#Posicion inicial del meteorito
#r0=np.array([-3.3921932,-5.3551800,3.76132566])*10**6
r0=np.array([0.5,-6.5,4.5])*10**6
nr0=math.sqrt(r0@r0)

#posicion del punto de impacto, the white house
rf=np.array([9.96554695, -0.56978729,  9.17335717])*10**6#array([1.1123098,-4.8319904,4.0005346])#*10**6
nrf=math.sqrt(rf@rf)
nr0

7921489.75887743

Ahora utilizamos los vectores anteriores para calcular la velocidad inicial como

$$
\vec{v}_0=\frac{\vec{r}_f-\vec{r}_0}{|\vec{r}_f-\vec{r}_0|} \times 1000 \ m/s
$$

In [2]:
resta=rf-r0
nresta=math.sqrt(resta@resta)
resta=resta/nresta 

#Velocidad inicial del meteorito
v0=resta*1366.85
nv0=math.sqrt(v0@v0)

De lo anterior obtenemos que la velocidad inicial resulta ser

$$
\vec{v}_0=(\ 1068.5461\ ,\ 669.44941\ ,\ 527.66566\ ) \ \ \ [\ m/s \ ] 
$$

Con ayuda de los vectores $\ \vec{r}_0 \ $ y $\ \vec{r}_f \ $ calculamos los vectores $\ \vec{i}\ '\ $ , $\ \vec{j}\ '\ $ , $\ \vec{k}\ '\ $ que conformaran la nueva base

In [3]:
#Vector unitario i'
i=r0/nr0

u=np.cross(i,rf/nrf)
nu=math.sqrt(u@u)

#vector unitario k'
k=u/nu

#Vector unitario j'
j=np.cross(k,i)

print('i\'=',i)
print('j\'=',j)
print('k\'=',k)

i'= [ 0.06311944 -0.82055272  0.56807496]
j'= [0.79728396 0.38383544 0.46584186]
k'= [-0.60029511  0.42351338  0.678441  ]


Ahora que tenemos los vectores calculamos la matriz de transformacion haciendo

\begin{equation}
\mathbb M=
\begin{pmatrix}
| & | & | \\
\vec{i}\ ' & \vec{j}\ ' & \vec{k}\ '\\
| & | & | \\
\end{pmatrix}
\end{equation}

In [4]:
#Matriz de cambio de base
M=np.array([i,j,k])
M=np.transpose(M)
print('M=')
print(M)

M=
[[ 0.06311944  0.79728396 -0.60029511]
 [-0.82055272  0.38383544  0.42351338]
 [ 0.56807496  0.46584186  0.678441  ]]


Ahora que tenemos la matriz de cambio de base, transformamos nuestras condiciones iniciales a la nueva base

$$
\vec{r}_0\ '=\mathbb M \cdot \vec{r}_0
$$

$$
\vec{v}_0\ '=\mathbb M \cdot \vec{v}_0
$$



In [5]:
r0t=np.linalg.solve(M,r0)
nr0t=math.sqrt(r0t@r0t)

v0t=np.linalg.solve(M,v0)
nv0t=math.sqrt(v0t@v0t)
print('r0\'=',r0t)
print(            )
print('v0\'=',v0t)

r0'= [ 7.92148976e+06 -3.06582673e-10 -3.97609917e-10]

v0'= [-1.82175732e+02  1.35465528e+03 -1.26153906e-13]


Los vectores en la nueva base son

$$
\vec{r}_0\ '=(\ 7.92148976\ ,\ 0\ ,\ 0\ ) \ \ \ [\ m\ \times\ 10^6\ ]
$$

$$
\vec{v}_0\ '=(\ -182.17573\ ,\ 1,354.6553\ ,\ 0\ ) \ \ \ [\ m/s \ ] 
$$

_____________________________________________________________________

Calculamos las magnitudes de estos vectores en la nueva base y las denotamos con "$r$" y "$v$"

In [6]:
r=nr0t
v=nv0t
print('r\'=',r/1e6,'x10^6 metros')
print( )
print('v\'=',v,'m/s')

r'= 7.9214897588774305 x10^6 metros

v'= 1366.8500000000001 m/s


Ahora definimos parametros importantes tales como

$$
m=5,000\ kg \ \ , \ \ GM=3.9861352\times 10^{14}\ m^2/s^3 
$$ 

Y algunas otras tales como

$$
\vec{L}= \vec{r} \times \vec{p} \ \ \ \ , \ \ \  \ l=L_z \ \ \ \ , \ \ \ \ E=\frac{1}{2}mv^2+\frac{l^2}{2mr^2}-\frac{mGM}{r} 
$$

In [7]:
# Establecemos las constantes
m=5000.
GM=3.9861352e14
Rt=6.371e6

# Calculamos L y l
L=m*np.cross(r0t,v0t)
l=math.sqrt(L@L)

# Calculamos la energia
E=0.5*m*v**2+l**2/(2*m*r**2)-m*GM/r

Con todas estas magnigtudes fisicas podemos calcular la ecuacion de la orbita que viene dada por

$$
r=\frac{\frac{l^2}{m^2GM}}{1+e\ cos(\phi-\phi')} \ \ \ \ e=\sqrt{1+\frac{2E}{m^3}\left(\frac{l}{GM}\right)^2}
$$

Vease la siguiente tabla de valores que relaciona la exentricidad con el tipo de orbita

$$
e=0 \ \ \ \ circunferencia
$$ 
$$
e< 1 \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \  \ \ \ \ elipse
$$ 
$$
e=1 \ \ \ \ \  \ \  \ \  \ \ \ \ \ \ \ parabola
$$ 
$$
e>1 \ \  \ \ \ \ \ \ \ \ \ \ \ \ \ hiperbola
$$ 

In [8]:
#Calculemos la excentricidad
e=math.sqrt(1+2*E*l**2/(m**3*GM**2))
e

0.9642343111981878

In [9]:
# Calculemos el termino del numerador
A=l**2/(GM*m**2)
A

288881.208938698

Con estas cantidades podemos concluir que la ecuacion de la orbita es

$$
r=\frac{28.88121}{1+0.96423431cos \phi} \ \ \ [\ km\ ]
$$

Ahora calculamos las constantes a, b y c

$$
a=-\frac{mGM}{2E} \ \ \ \ \ , \ \ \ \ \ b=\frac{l}{\sqrt{-2mE}} \ \ \ \ \ , \ \ \ \ \ c=\sqrt{a^2-b^2}
$$



In [10]:
a=-m*GM/(2*E)
b=l/math.sqrt(-2*m*E)
c=math.sqrt(a**2-b**2)

a/1e6,b/1e6,c/1e6

(4.112059654763421, 1.0899067686256938, 3.964989008816665)

Su equivalente cartesiana es

$$
\frac{(x-3.9649890)^2}{4.11205965^2}+\frac{(y-0)^2}{1.0899067^2}=1
$$


Lo anterior en cartesianas:

$$
x'=a\ cos\phi \ \ \ \ \ , \ \ \ \ \ y'=b\ sin\phi
$$



Ahora evaluaremos la ecuacion en $\ r_f\ $ para encontrar $\ \phi_f\ $  

In [11]:
phi=-math.acos((A/Rt-1)/e)
print(phi*180/np.pi)

-171.91776126801173


Ahora calculamos las coordenadas $\ x_f'\ $ y $\ y_f'\ $ que componen al vector $\vec{r}_f\ '$ o coordenadas del punto de impacto referidas a la base primada

In [12]:
u2=2*math.atan(math.sqrt((1-e)/(1+e))*math.tan(phi/2))
#u2=math.acos(e+6.371e6*math.cos(phi)/a)
xt=Rt*math.cos(np.pi-phi)
yt=-Rt*math.sin(np.pi-phi)

rft=np.array([xt,yt,0.])
rft[0]="{0:.1f}".format(rft[0])
rft[1]="{0:.2f}".format(rft[1])
print('rf\'=',rft)
print(math.sqrt(rft@rft)/1e6)

rf'= [6307718.7   895726.45       0.  ]
6.371000005616017


Ahora que tenemos el vector $\ \vec{r}_f\ '\ $ calculamos su equivalente en la base de referencia mediante

$$
\vec{r}_f=\mathbb M ^{-1} \cdot \vec{r}_f\ '
$$

In [13]:
N=np.linalg.inv(M)

rf2=np.linalg.solve(N,rft)
nrf2=math.sqrt(rf2@rf2)

Pasamos estas coordenadas de cartesianas a terrestres


In [14]:
x__=rf2[0] ; y__=rf2[1] ; z__=rf2[2]

if z__>0:
	theta__=math.atan(math.sqrt(x__**2+y__**2)/z__)

if (z__<0.):
	theta__=np.pi+math.atan(math.sqrt(x__**2+y__**2)/z__)


if (x__>0. and y__>0.):
	phi__=math.atan(y__/x__)

if (x__>0. and y__<0.):
	phi__=2*np.pi+math.atan(y__/x__)

if (x__<0. and y__>0. or x__<0. and y__<0):
	phi__=np.pi+math.atan(y__/x__)

phi__=phi__*180/np.pi
theta__=theta__*180/np.pi

print('phi=',phi__,'theta=',theta__)

phi= 282.9632042162626 theta= 51.10264277638966


In [15]:
if (phi__>=0. and phi__<=180.):
	long__=phi__

if (phi__>180. and phi__<360):
	long__=360-phi__
	long__=-long__

 
if (theta__>0. and theta__<90):
	lat__=90.-theta__

if (theta__>270. and theta__<360.):
	lat__=theta__-90.
	lat__=-lat__

print(''+str("{0:.6f}".format(lat__))+',',"{0:.6f}".format(long__))


38.897357, -77.036796
